File: frontend/src/modules/app/api/calendar/entities/appointment.ts
import { AppointmentBlock } from '@/modules/app/api/calendar/entities/appointment-block';
import { Entity } from '@/core';
import moment from 'moment';

export class Appointment extends Entity {
    public static NOTES_MAX_LENGTH: number = 1024;

    /**
     * Create a new appointment.
     * @param serviceId The id of the service being performed.
     * @param price The price in dollars.
     * @param duration The minutes it will take.
     * @param blocks The times the appointment is on.
     * @param notes Misc. notes.
     */
    constructor(
        public serviceId: string,
        public clientId: string,
        public price: number,
        public blocks: AppointmentBlock[],
        public notes?: string
    ) {
        super();

        if (this.notes != null && this.notes.length > Appointment.NOTES_MAX_LENGTH) {
            throw new RangeError();
        }
    }

    get duration(): number {
        return this.blocks.map(b => b.duration).reduce((a, b) => a + b, 0);
    }

    getBlocksForDay(date: Date): AppointmentBlock[] {
        const dayM = moment(date);

        return this.blocks.filter(b => {
            const blockM = moment(b.date);
            return blockM.isSame(dayM, 'day');
        });
    }

    getBlocksForWeek(date: Date): AppointmentBlock[] {
        const weekM = moment(date);

        return this.blocks.filter(b => {
            const blockM = moment(b.date);
            return blockM.isSame(weekM, 'week');
        });
    }
}

File: api/src/Domain/Calendar/Entities/Appointment.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DetailingArsenal.Domain {
    public class Appointment : Aggregate<Appointment>, IUserEntity {
        public const int NotesMaxLength = 1024;

        public Guid UserId { get; set; }

        /// <summary>
        /// Id of the service being performed.
        /// </summary>
        public Guid ServiceId { get; set; }

        public Guid ClientId { get; set; }

        /// <summary>
        /// Price of the appointment in USD.
        /// </summary>
        public decimal Price { get; set; }

        /// <summary>
        /// How long the appointment will take in minutes.
        /// </summary>
        public int Duration => Blocks.Select(b => b.Duration).Sum();

        /// <summary>
        /// Misc. notes.
        /// </summary>
        public string? Notes {
            get => notes;
            set {
                if (notes?.Length > NotesMaxLength) {
                    throw new ArgumentOutOfRangeException();
                }

                notes = value;
            }
        }

        public List<AppointmentBlock> Blocks { get; set; } = new List<AppointmentBlock>();

        private string? notes;
    }
}